testkit/src/main/scala/ch/epfl/scala/bsp/testkit/mock/HappyMockServer.scala

```diff
@@ -71,223 +106,388 @@ class HappyMockServer(base: File) extends AbstractMockServer {
+        // TODO some individual test notifications
+        val result = new TestResult(StatusCode.OK)
+        result.setOriginId(params.getOriginId)
+        Left(result)
+      }
+    })
+
+  override def buildTargetRun(params: RunParams): CompletableFuture[RunResult] =
+    handleRequest(() => {
+      compileTargets.get(params.getTarget) match {
+        case Some(target) =>
+          if (!target.getCapabilities.getCanRun)
+            Right(
+              new ResponseError(
+                ResponseErrorCode.InvalidParams,
+                s"Target ${target.getId.getUri} is not compilable",
+                null
+              )
+            )
+          else
+            mockRunAnswer(params)
+        case None =>
+          mockRunAnswer(params)
+      }
+    })
+
+  private def mockRunAnswer(params: RunParams) = {
+    // TODO some task notifications
+    val result = new RunResult(StatusCode.OK)
     result.setOriginId(params.getOriginId)
-    completeFuture(result)
+    Left(result)
   }
 
-  override def buildTargetTest(params: TestParams): CompletableFuture[TestResult] = {
-    // TODO some test task/report notifications
-    // TODO some individual test notifications
-    val result = new TestResult(StatusCode.OK)
-    result.setOriginId(params.getOriginId)
-    completeFuture(result)
+  override def buildTargetCleanCache(
+      params: CleanCacheParams
+  ): CompletableFuture[CleanCacheResult] =
+    handleRequest(() => {
+      val result = new CleanCacheResult("cleaned cache", true)
+      Left(result)
+    })
+
+  private def handleRequest[T](
+      f: () => Either[T, ResponseError],
+      isBuildInitialize: Boolean = false,
+      isBuildShutdown: Boolean = false
+  ): CompletableFuture[T] = {
+    val future = CompletableFuture.supplyAsync(f.asJava)
+    if (isShutdown.isCompleted && !isBuildShutdown)
+      completeExceptionally(
+        new ResponseError(
+          ResponseErrorCode.serverErrorEnd,
+          "Cannot handle requests after receiving the shutdown request",
+          null
+        )
+      )
+    else if (!isInitialized.isCompleted && !isBuildInitialize) {
+      Try(Await.result(isInitialized.future, 1.seconds)) match {
+        case Failure(_) =>
+          completeExceptionally(
+            new ResponseError(
+              ResponseErrorCode.serverNotInitialized,
+              "Cannot handle requests before receiving the initialize request",
+              null
+            )
+          )
+        case Success(_) => getValue(future)
+      }
+    } else {
+      getValue(future)
+    }
   }
 
-  override def buildTargetRun(params: RunParams): CompletableFuture[RunResult] = {
-    // TODO some task notifications
-    val result = new RunResult(StatusCode.OK)
-    result.setOriginId(params.getOriginId)
-    completeFuture(result)
+  private def getValue[T](future: CompletableFuture[Either[T, ResponseError]]): CompletableFuture[T] = {
+    Try(future.get()) match {
```

Comment:
function returns a CompletableFuture, so waiting here is not necessary, you can chain instead.